### PR TITLE
Position Store Refactor

### PR DESCRIPTION
**Pull request by ming535**
1. Refactor position store.
  
    - Use position repo and position cache instead of one big position store.

    - Initialization logic and serialization logic for position is owned by input plugin.

    - Make it clear that position cache's lifecycle is managed by server, not by input plugin.

    - Mongo oplog position use new format.

    - Kafka offset position use new format.

2. Do not do duplicate scan when mysql scan is done.

### CODE REVIEW

File: pkg/inputs/mysqlbatch/mysql_table_scanner.go [hunk 1]
```diff
@@ -371,7 +371,7 @@ func (tableScanner *TableScanner) FindAll(db *sql.DB, tableDef *schema_store.Tab
 
 	for i := range allData {
 		rowPtrs := allData[i]
-		msg := NewMsg(rowPtrs, columnTypes, tableDef, nil, position_store.MySQLTablePosition{})
+		msg := NewMsg(rowPtrs, columnTypes, tableDef, nil, position_store.MySQLTablePosition{Column: "*", Type: position_store.PlainInt, Value: i})
```
Comment (Ryan-Git) [inline pkg/inputs/mysqlbatch/mysql_table_scanner.go]: how can this position help?

File: cmd/gravity/main.go [hunk 2]
```diff
@@ -165,6 +164,7 @@ func main() {
 
 func healthzHandler(server *app.Server) func(http.ResponseWriter, *http.Request) {
 	return func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
```
Comment (Ryan-Git) [inline cmd/gravity/main.go]: this ignores health check, why?

File: pkg/inputs/helper/mysql_common.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+func SerializeBinlogPositionValue(position *BinlogPositionsValue) (string, error) {
+	s, err := myJson.MarshalToString(position)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return s, nil
+}
+
+func DeserializeBinlogPositionValue(value string) (*BinlogPositionsValue, error) {
```
Comment (Ryan-Git) [inline pkg/inputs/helper/mysql_common.go:32]: panic on error is easier I think. Caller can do nothing than exist

File: pkg/core/input.go [hunk 4]
```diff
@@ -6,14 +6,11 @@ import (
...
+	NewPositionCache() (position_store.PositionCacheInterface, error)
+	Start(emitter Emitter, router Router, positionCache position_store.PositionCacheInterface) error
 	Close()
 	Stage() config.InputMode
-	// TODO position store can be hidden by input plugin
-	// or we should use a configuration dedicated for position store
-	NewPositionStore() (position_store.PositionStore, error)
-	PositionStore() position_store.PositionStore
-	Done() chan position_store.Position
+	Done(positionCache position_store.PositionCacheInterface) chan position_store.Position
```
Comment (Ryan-Git) [inline pkg/core/input.go]: why `Done` need `positionCache`? `Input` should always have a ref to it.

File: pkg/core/input.go [hunk 5]
```diff
@@ -6,14 +6,11 @@ import (
 )
 
 type Input interface {
-	Start(emitter Emitter, router Router) error
+	NewPositionCache() (position_store.PositionCacheInterface, error)
+	Start(emitter Emitter, router Router, positionCache position_store.PositionCacheInterface) error
```
Comment (Ryan-Git) [inline pkg/core/input.go:9]: this seems strange. If `Input` is the only component able to create position cache, then `Start` should not inject it.
The other way is, `PositionCache` is created by `server`, and injected into `Input`, then there shouldn't have `NewPositionCache` method.

**Review by Ryan-Git — CHANGES_REQUESTED** [review]
why each `Input` has its own mutex and logic to get/update position value? I think just serialize 
 & deserialize are different.

File: cmd/gravity/main.go (hunk 2 above)
Comment (ming535) [reply cmd/gravity/main.go]: fixed

File: pkg/inputs/helper/mysql_common.go (hunk 3 above)
Comment (ming535) [reply pkg/inputs/helper/mysql_common.go:32]: it's better for the caller to exit, so that we can know the caller's stack trace when error happens

Comment (Ryan-Git) [reply pkg/inputs/helper/mysql_common.go:32]: if we do a panic rather than fatal, the stack is also kept.

**Review by Ryan-Git — APPROVED** [review]

**Pull request merged**